ambry-router/src/main/java/com/github/ambry/router/CompressionMetrics.java

```diff
@@ -0,0 +1,141 @@
+    public final Meter decompressRate;     // Rate decompress() is called.
+    public final Counter decompressError;  // Rate decompress() throws exception.
+    public final Histogram fullSizeDecompressTimeInMicroseconds;
+    public final Histogram fullSizeDecompressSpeedMBPerSec;
+    public final Histogram partialFullDecompressTimeInMicroseconds;
+    public final Histogram partialFullDecompressSpeedMBPerSec;
+
+    /**
+     * Create an instance of algorithm-specific metrics.
+     * @param algorithmName Name of the algorithm.  The metric name will contain the algorithm name.
+     * @param registry The metric registry to add the new metrics.
+     */
+    public AlgorithmMetrics(String algorithmName, MetricRegistry registry) {
+      // Compression metrics
+      compressRate = registry.meter(MetricRegistry.name(ownerClass, algorithmName, "CompressRate"));
+      compressError = registry.counter(MetricRegistry.name(ownerClass, algorithmName, "CompressError"));
+      compressRatioPercent = registry.histogram(MetricRegistry.name(ownerClass, algorithmName, "CompressRatioPercent"));
+      fullSizeCompressTimeInMicroseconds = registry.histogram(MetricRegistry.name(ownerClass, algorithmName,
+          "FullSizeCompressTimeInMicroseconds"));
+      fullSizeCompressSpeedMBPerSec = registry.histogram(MetricRegistry.name(ownerClass, algorithmName,
+          "FullSizeCompressSpeedMBPerSec"));
+      partialFullCompressTimeInMicroseconds = registry.histogram(MetricRegistry.name(ownerClass, algorithmName,
+          "PartialFullCompressTimeInMicroseconds"));
+      partialFullCompressSpeedMBPerSec = registry.histogram(MetricRegistry.name(ownerClass, algorithmName,
+          "PartialFullCompressSpeedMBPerSec"));
+
+      // Decompression metrics.
+      decompressRate = registry.meter(MetricRegistry.name(ownerClass, algorithmName, "DecompressRate"));
+      decompressError = registry.counter(MetricRegistry.name(ownerClass, algorithmName, "DecompressError"));
+      fullSizeDecompressTimeInMicroseconds = registry.histogram(MetricRegistry.name(ownerClass, algorithmName,
+          "FullSizeDecompressTimeInMicroseconds"));
+      fullSizeDecompressSpeedMBPerSec = registry.histogram(MetricRegistry.name(ownerClass, algorithmName,
+          "FullSizeDecompressSpeedMBPerSec"));
+      partialFullDecompressTimeInMicroseconds = registry.histogram(MetricRegistry.name(ownerClass, algorithmName,
+          "PartialFullDecompressTimeInMicroseconds"));
+      partialFullDecompressSpeedMBPerSec = registry.histogram(MetricRegistry.name(ownerClass, algorithmName,
+          "PartialFullDecompressSpeedMBPerSec"));
+    }
+  }
+
+  // Group all compression related metrics under one owner "Compression".
+  // These metrics are per chunk, not per operation.
+  private final static Class<?> ownerClass = Compression.class;
+  private final MetricRegistry metricsRegistry;
+  private final Map<String, AlgorithmMetrics> algorithmToMetricsMap = new ConcurrentHashMap<>();
```

Comment:
are there going to be multiple algorithms operating at the same time ?

Reply:
There will be only 1 selected compression algorithm, but decompression may use different ones.  Yes, we do support multiple concurrent decompression algorithms.  That's why we store the algorithm name inside the compressed chunk.

Reply:
hmm so if a blob is compressed using algo X, can it be decompressed using algo Y ? 
or do u mean blob1 is compressed/decompressed using algo X, blob2 using algo Y and blob3 using algo Z ?

Reply:
If a blob is compressed using algo X, it must be decompressed by algo X.  However, if a new or better algo is available, we would want to switch to the new algo.  However, previously compressed blobs still must be decompressed using original compressor.